File: leetcode/editor/en/82.remove-duplicates-from-sorted-list-ii/82.remove-duplicates-from-sorted-list-ii.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */


class ListNode {
    val: number
    next: ListNode | null

    constructor(val?: number, next?: ListNode | null) {
        this.val = (val === undefined ? 0 : val)
        this.next = (next === undefined ? null : next)
    }
}

function deleteDuplicates(head: ListNode | null): ListNode | null {
    if (head === null || head.next === null) return head

    const dummy = new ListNode(1e10)
    dummy.next = head
    let p: ListNode | null = dummy

    while (p !== null) {
        while (p.next !== null) {
            const tempNext: ListNode = p.next

            let q: ListNode | null = tempNext.next
            while (q !== null && q.val === tempNext.val)
                q = q.next

            if (tempNext.next !== q)
                p.next = q
            else
                break
        }

        p = p.next
    }

    return dummy.next
}

let head = new ListNode(1)
head.next = new ListNode(2)
head.next.next = new ListNode(4)
head.next.next.next = new ListNode(4)
head.next.next.next.next = new ListNode(5)
deleteDuplicates(head)

File: leetcode/editor/en/82.remove-duplicates-from-sorted-list-ii/82.remove-duplicates-from-sorted-list-ii.py
# //////////# # Given the head of a sorted linked list, delete all nodes that 
# //have 
# //////////# duplicate 
# //////////# # numbers, leaving only distinct numbers from the original list. 
# ////Return 
# //////
# ////////the 
# //////////
# //////////# linked 
# //////////# # list sorted as well. 
# //////////# # 
# //////////# # 
# //////////# # Example 1: 
# //////////# # 
# //////////# # 
# //////////# # Input: head = [1,2,3,3,4,4,5]
# //////////# # Output: [1,2,5]
# //////////# # 
# //////////# # 
# //////////# # Example 2: 
# //////////# # 
# //////////# # 
# //////////# # Input: head = [1,1,1,2,3]
# //////////# # Output: [2,3]
# //////////# # 
# //////////# # 
# //////////# # 
# //////////# # Constraints: 
# //////////# # 
# //////////# # 
# //////////# # The number of nodes in the list is in the range [0, 300]. 
# //////////# # -100 <= Node.val <= 100 
# //////////# # The list is guaranteed to be sorted in ascending order. 
# //////////# # 
# //////////# # Related Topics Linked List Two Pointers 👍 4797 👎 153
# //////////# 
# //////////
# ////////
# //////
# ////
# //
# 


# leetcode submit region begin(Prohibit modification and deletion)
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def deleteDuplicates(self, head: Optional[ListNode]) -> Optional[ListNode]:
        
# leetcode submit region end(Prohibit modification and deletion)
